api/lql.go

```diff
@@ -18,23 +18,115 @@
 
 package api
 
+import (
+	"encoding/json"
+	"net/url"
+	"regexp"
+)
+
+const (
+	reLQL string = `(?ms)^(\w+)\(\w+\s\w+\)\s*{`
+)
+
+type LQLQuery struct {
+	ID             string `json:"LQL_ID,omitempty"`
+	StartTimeRange string `json:"START_TIME_RANGE,omitempty"`
+	EndTimeRange   string `json:"END_TIME_RANGE,omitempty"`
+	QueryText      string `json:"QUERY_TEXT"`
+	// QueryBlob is a special string that supports type conversion
+	// back and forth from LQL to JSON
+	QueryBlob string `json:"-"`
+}
+
+func (q *LQLQuery) translate() {
+	// if QueryText is not already populated
+	if q.QueryText == "" {
+		// if QueryBlob is JSON
+		var t LQLQuery
+		err := json.Unmarshal([]byte(q.QueryBlob), &t)
+
+		if err == nil {
+			q.StartTimeRange = t.StartTimeRange
+			q.EndTimeRange = t.EndTimeRange
+			q.QueryText = t.QueryText
+			// if QueryBlob is LQL
+		} else if matched, _ := regexp.MatchString(reLQL, q.QueryBlob); matched {
+			q.QueryText = q.QueryBlob
+		}
+	}
+}
+
+type LQLQueryResponse struct {
+	Data    []LQLQuery `json:"data"`
+	Ok      bool       `json:"ok"`
+	Message string     `json:"message"`
+}
+
 // LQLService is a service that interacts with the LQL
 // endpoints from the Lacework Server
 type LQLService struct {
 	client *Client
 }
 
-type lqlQueryRequest struct {
-	QueryText string `json:"QUERY_TEXT"`
+func (svc *LQLService) CreateQuery(query string) (
+	response LQLQueryResponse,
+	err error,
+) {
+	lqlQuery := LQLQuery{QueryBlob: query}
+	lqlQuery.translate()
```

Comment:
What happens if the `LQLQuery` struct is malformed? Meaning that if someone doesn't put any Query, should we throw an error inside the `translate()` function? 🤔

Reply:
> What happens if the LQLQuery struct is malformed?

In this case however the underlying api deals with an empty `QUERY_TEXT` value, but we can add error checking in the `translate()` function if so desired...

Reply:
Fixed.